htlcswitch/htlcnotifier.go

```diff
@@ -0,0 +1,381 @@
+type HtlcEventType int
+
+const (
+	// HtlcEventTypeSend represents a htlc that was part of a send from
+	// our node.
+	HtlcEventTypeSend HtlcEventType = iota
+
+	// HtlcEventTypeReceive represents a htlc that was part of a receive
+	// to our node.
+	HtlcEventTypeReceive
+
+	// HtlcEventTypeForward represents a htlc that was forwarded through
+	// our node, which was not part of a send or receive.
+	HtlcEventTypeForward
+)
+
+// ForwardingEvent represents a htlc that was forwarded onwards from our node.
+// Sends which originate from our node will report forward events with zero
+// incoming channel IDs in their HtlcKey.
+type ForwardingEvent struct {
+	// HtlcKey contains the information which uniquely identifies the htlc.
+	HtlcKey
+
+	// HtlcInfo contains further information about the htlc.
+	HtlcInfo
+
+	// Timestamp is the time when this htlc was forwarded.
+	Timestamp time.Time
+}
+
+// LinkFailEvent describes a htlc that failed on our incoming our outgoing
+// link. The incoming bool is true for failures on incoming links, and false
+// for failures on outgoing links. The failure reason is provided by a lnwire
+// failure message which is enriched with a failure detail in the cases where
+// the wire failure message does not discolse full information about the
+// failure.
+type LinkFailEvent struct {
+	// HtlcKey contains the information which uniquely identifies the htlc.
+	HtlcKey
+
+	// HtlcInfo contains further information about the htlc.
+	HtlcInfo
+
+	// LinkError is the reason that we failed the htlc.
+	LinkError *LinkError
+
+	// Incoming is true if the htlc was failed on an incoming link.
+	// If it failed on the outgoing link, it is false.
+	Incoming bool
+
+	// Timestamp is the time when the link failure occurred.
+	Timestamp time.Time
+}
+
+// ForwardingFailEvent represents a htlc failure which occurred down the line
+// after we forwarded a htlc onwards. An error is not included in this event
+// because errors returned down the route are encrypted.
+type ForwardingFailEvent struct {
+	// HtlcKey contains the information which uniquely identifies the htlc.
+	HtlcKey
+
+	// HtlcInfo contains further information about the htlc.
+	HtlcInfo
+
+	// Timestamp is the time when the forwarding failure was received.
+	Timestamp time.Time
+}
+
+// SettleEvent represents a htlc that we settled.
+type SettleEvent struct {
+	// HtlcKey contains the information which uniquely identifies the htlc.
+	HtlcKey
+
+	// HtlcInfo contains further information about the htlc.
+	HtlcInfo
+
+	// Timestamp is the time when this htlc was settled.
+	Timestamp time.Time
+}
+
+// NotifyForwardingEvent notifies the HtlcNotifier than a htlc has been
+// forwarded.
+//
+// Note this is part of the htlcNotifier interface.
+func (h *HtlcNotifier) NotifyForwardingEvent(pkt *htlcPacket,
+	hash lntypes.Hash) {
+
+	event := &ForwardingEvent{
+		HtlcKey:   newHtlcKey(pkt),
+		HtlcInfo:  newHtlcInfo(pkt, hash, getEventType(pkt)),
+		Timestamp: time.Now(),
+	}
```

Comment:
I am not totally fully comfortable with the function signature and passing in `htlcPacket`. Especially because sometimes an instance of `htlcPacket` is created only for the notification. Isn't it better to let the caller instantiate the event? Then the `HtlcNotifier` could maybe even disappear completely and replaced by just an instance of `subscribe.Server`.

Reply:
Decided on passing in raw packets because there's some `hltcNotifier` side logic that I don't want calling functions to have to do (eg, figuring out event type based on presence of channel IDs). Could do this with a constructor for each type, it will just have a ton of arguments?

idk about just a subscribe server, our current pattern is to wrap these servers in structs which strictly control what can be piped in which I like. Plus subscribe servers aren't great for unit tests, whereas htlcnotifier can unobtrusively mocked when we need.

Reply:
I think it is much cleaner now that you keep the packet in the link/switch.